_posts/2020-04-11-native-mobile-to-flutter.markdown

```diff
@@ -0,0 +1,122 @@
+- Flutter does not rely on any native UI components, so it's largely immune to breaking changes introduced by Apple, Google, or Android device manufacturers.
+
+### Choosing a State Management pattern that matched our needs.
+
+There were several state management architectures to choose from, but we opted for simplicity, the [Scoped Model pattern](https://github.com/brianegan/scoped_model).  The scoped model pattern is a simple change notifier pattern bound to Flutter's UI widget hierarchy.  It has gradually been superceded in the community by the [Provider library](https://github.com/rrousselGit/provide) which is based on Scoped Model.  We will most likely be migrating to the provider library within the next year.
+
+We use this pattern with our SQLite database to allow sql table subscriptions that trigger redraws on subscribed user interfaces when data is modified on disk.  We also use a very simple stream subscription pattern to propagate state changes that are not tied to database state.  Flutter provides the excellent [StreamBuilder class](https://api.flutter.dev/flutter/widgets/StreamBuilder-class.html) to facilitate state propagation.
+
+We evaluated a few other patterns mentioned in flutter resources, particularly Redux and BLoC, but chose the simplicity of Scoped Model over the granularity and associated complexity that the others presented.  
+
+### Native integrations. 
+
+Flutter offers a very simple message passing api for communicating with native code.  We needed to use native integrations for a relatively small group of features: 
+
+- Push notification.
+- Deep link handling.
+- Connectivity detection.
+- Security features.
+- Device information. 
+
+_This platform specific code amounted to less than 2% of our total lines of code._ 
+
+We were able to write this functionality easily and communicate between Flutter and native with very little code overhead. Flutter provides a [very simple api](https://flutter.dev/docs/development/platform-integration/platform-channels#architecture) for communicating bidirectionally between native and dart code.
+
+### Unit tests.
+
+Unit testing in Flutter is straightforward, but lacks the robust testing utilities and third party libraries found in React.  We rolled most of our own testing scaffolding aside from the simple test runner offered by flutter.
+
+### UI testing. 
+
+UI testing in Flutter only runs in a simulator/emulator, a pretty significant drawback.  Though we were able to write effective widget tests, we found them to be brittle.  This is definitely an area that needs improvements in future iterations of Flutter.  We opted to keep the majority of presentation logic in view models so that we could test the presentation layer with unit tests wherever possible rather than employing widget tests.
+
+### Continuous integration.
```

Comment:
Just thinking a loud as I'm reading this, but it might be cool to split CI, deployment, release cycle etc into a separate blog post. That would allow you to go into more detail (e.g. what does CodeMagic do specifically that's so great?)

Up to this point, the thrust of the article has been React vs Flutter => state management so might be nice to keep the article centered around that.

Reply:
Maybe adding a little context to what we're trying to convey.  I think its kind of a good pro-tip for devs who want to get things up and running fast.  We don't really have a CI preference, codemagic was the only turnkey solution we found.  It was the only solution that offered us that ability out of the box and allowed us to do the entire setup via GUI.  We're not trying to advise what to do, here, but rather what worked for us.  If someone is looking to add CI enforcement or build pipelines, they can use this instead of investing man hours.  At this point, there is no other turnkey solution, but as time passes, all of the other players will probably offer them.

Reply:
That's a great point. I agree that it would be stronger if keeping the focus on the choice of Flutter over RN. Also, I'd enjoy reading more in depth in a separate post about choices the team has made on the CI and release cycle side. :)